umh-core/pkg/service/filesystem/buffered_filesystem.go

```diff
@@ -0,0 +1,786 @@
+	bs.changed[path] = &fileChange{removed: true, wasDir: targetIsDir}
+	return nil
+}
+
+// Stat returns an os.FileInfo-like object if it is in the in-memory map. Otherwise os.ErrNotExist.
+func (bs *BufferedService) Stat(ctx context.Context, path string) (os.FileInfo, error) {
+	bs.mu.Lock()
+	defer bs.mu.Unlock()
+
+	st, ok := bs.files[path]
+	if !ok {
+		return nil, os.ErrNotExist
+	}
+	if st.isDir {
+		// Return a synthetic fileInfo for a directory
+		return &memFileInfo{
+			name:  filepathBase(path),
+			size:  0,
+			mode:  st.fileMode,
+			mtime: st.modTime,
+			dir:   true,
+		}, nil
+	}
+	return &memFileInfo{
+		name:  filepathBase(path),
+		size:  st.size,
+		mode:  st.fileMode,
+		mtime: st.modTime,
+		dir:   false,
+	}, nil
+}
+
+// CreateFile creates a file in-memory and marks it for creation on disk during SyncToDisk.
+// It returns nil, nil because the actual file doesn't exist on disk yet.
+func (bs *BufferedService) CreateFile(ctx context.Context, path string, perm os.FileMode) (*os.File, error) {
+	bs.mu.Lock()
+	defer bs.mu.Unlock()
+
+	// If already exists, check if it is a file
+	if state, ok := bs.files[path]; ok {
+		if state.isDir {
+			return nil, fmt.Errorf("%s exists and is not a file", path)
+		}
+		// It exists and is a file, so we're good
+		return nil, nil
+	}
+
+	// Create new file entry in memory
+	bs.files[path] = &fileState{
+		isDir:    false,
+		content:  []byte{},
+		modTime:  time.Now(),
+		fileMode: perm,
+		size:     0,
+	}
+
+	// Mark as changed so that SyncToDisk will create it on disk
+	bs.changed[path] = &fileChange{
+		content: []byte{},
+		perm:    perm,
+		removed: false,
+	}
+
+	// Return nil, nil since the file isn't actually on disk yet
+	return nil, nil
+}
+
+// Chmod updates the fileMode in memory (and later flush).
+func (bs *BufferedService) Chmod(ctx context.Context, path string, mode os.FileMode) error {
+	bs.mu.Lock()
+	defer bs.mu.Unlock()
+
+	st, ok := bs.files[path]
+	if !ok {
+		return os.ErrNotExist
+	}
+	st.fileMode = mode
+
+	// If the file wasn't removed, mark changed
+	chg, inChg := bs.changed[path]
+	if !inChg {
+		chg = &fileChange{}
+	}
+	chg.perm = mode
+	bs.changed[path] = chg
+
+	return nil
+}
+
+// ReadDir returns directories only from the in-memory map.
+func (bs *BufferedService) ReadDir(ctx context.Context, path string) ([]os.DirEntry, error) {
+	bs.mu.Lock()
+	defer bs.mu.Unlock()
+
+	// Check if it's a directory in memory
+	st, ok := bs.files[path]
+	if !ok || !st.isDir {
+		return nil, fmt.Errorf("not a directory: %s", path)
+	}
+
+	// Gather immediate children (1 level)
+	var entries []os.DirEntry
+	prefix := path
+	if prefix != "" && !strings.HasSuffix(prefix, string(os.PathSeparator)) {
+		prefix += string(os.PathSeparator)
+	}
+
+	for filePath, fileState := range bs.files {
+		// If it has the prefix and does not contain another separator after prefix, it's an immediate child
+		if filePath != path && strings.HasPrefix(filePath, prefix) {
```

Comment:
Consider reducing nesting here, to improve readability

Reply:
added a code comment for this